### PR TITLE
reason per record, not per batch #147

### LINKED ISSUE #147: reason per record, not per batch
```
    def done(self, msg_ids: List[str], result: Optional[Result] = None, reason: Optional[str] = None):
```

### PR DESCRIPTION
**Pull request by spicy-sauce**

### CODE REVIEW

File: core/src/datayoga_core/block.py [hunk 1]
```diff
@@ -4,21 +4,19 @@
 import logging
 import os
 import sys
-from enum import Enum
+from abc import ABCMeta, abstractmethod
```
Comment (zalmane) [inline core/src/datayoga_core/block.py:7]: what is this ABCMeta?

File: core/src/datayoga_core/blocks/add_field/tests/test_add_field.py [hunk 2]
```diff
@@ -21,7 +22,7 @@ async def test_add_nested_field():
                   "expression": "[name.fname,name.lname] | join(' ', @)"})
     block.init()
     assert await block.run([{"name": {"fname": "john", "lname": "doe"}}]) == ([
-        {"name": {"fname": "john", "lname": "doe", "full_name": "john doe"}}], [Result.SUCCESS])
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/add_field/tests/test_add_field.py:24]: why the change from the Enum?

File: core/src/datayoga_core/blocks/add_field/tests/test_add_field.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
 import pytest
-from datayoga_core.block import Result
 from datayoga_core.blocks.add_field.block import Block
+from datayoga_core.result import Result
 
+RESULT_SUCCESS = Result.success()
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/add_field/tests/test_add_field.py]: why a method and not a plain enum?

File: core/src/datayoga_core/result.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        self.message = message
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, self.__class__) and
+            self.status == other.status and
+            self.message == other.message)
+
+    @staticmethod
+    def success() -> Result:
```
Comment (zalmane) [inline core/src/datayoga_core/result.py]: why a static method and not a contstant/enum?

File: core/src/datayoga_core/utils.py [hunk 5]
```diff
@@ -88,3 +90,17 @@ def get_connection_details(connection_name: str, context: Context) -> Dict[str,
             return connection
 
     raise ValueError(f"{connection_name} connection not found")
+
+
+def produce_data_and_results(data: List[Dict[str, Any]]) -> Tuple[List[Dict[str, Any]], List[Result]]:
```
Comment (zalmane) [inline core/src/datayoga_core/utils.py:96]: what is the purpose of this method?

File: integration-tests/test_stdin_to_stdout.py [hunk 6]
```diff
@@ -4,7 +4,7 @@
 from common.utils import run_job
 
 
-def test_stdin_to_stdout(tmpdir: str):
```
Comment (zalmane) [inline integration-tests/test_stdin_to_stdout.py:7]: any reason this was changed?

File: core/src/datayoga_core/block.py (hunk 1 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/block.py:7]: https://docs.python.org/3/library/abc.html

This module provides the infrastructure for defining [abstract base classes](https://docs.python.org/3/glossary.html#term-abstract-base-class) (ABCs) in Python

File: core/src/datayoga_core/blocks/add_field/tests/test_add_field.py (hunk 2 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/add_field/tests/test_add_field.py:24]: It's not Enum anymore, it's a Result object with SUCCESS status.

File: core/src/datayoga_core/blocks/add_field/tests/test_add_field.py (hunk 3 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/add_field/tests/test_add_field.py]: answered in a different comment

File: core/src/datayoga_core/result.py (hunk 4 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/result.py]: answered in a different comment

File: core/src/datayoga_core/utils.py (hunk 5 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/utils.py:96]: it goes over the records and puts the internal result field (Block.RESULT_FIELD) in the results array. If it's not specified, it's success. Then, it removes this field from the record.

File: integration-tests/test_stdin_to_stdout.py (hunk 6 above)
Comment (spicy-sauce) [reply integration-tests/test_stdin_to_stdout.py:7]: yes, it's not a str, but a pytest object and it was confusing to see read() on str.

File: core/src/datayoga_core/blocks/add_field/tests/test_add_field.py (hunk 2 above)
Comment (zalmane) [reply core/src/datayoga_core/blocks/add_field/tests/test_add_field.py:24]: Maybe use @dataclass unless we actually have business logic in those? This mode of a 'factory' singleton just seems unnatural. especially since we are creating one as a constant, but then duplicating it.

Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/add_field/tests/test_add_field.py:24]: From @zalmane :
Here is my suggestion. I would not mix static and non static methods in that dataclass. It is really meant as a Struct or NamedDict:
result.py
success= Result(Status.SUCCESS)
@dataclass
class Result():
    status: Status
    message: Optional[str] = None
  
    @staticmethod - remove this altogether. the code can just do this one liner and makes it more readable and easier to maintain
    def reject(message: Optional[str] = None) -> Result:
        return Result(Status.REJECTED, message)
so you use it:
import result
result.success

Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/add_field/tests/test_add_field.py:24]: done

**Review by zalmane — APPROVED** [review]
